File: sections/63_atcoder_difficulty_ladder/problems/41_median_pyramid_hard/solution.py
import sys


def main() -> None:
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    n = int(data[0])
    start = data[1]
    target = data[2]

    previous_operation = 0
    answer = 0
    for index in range(n - 1):
        mismatch = start[index] != target[index]
        current_operation = int(mismatch) ^ previous_operation
        answer += current_operation
        previous_operation = current_operation

    if previous_operation != (start[-1] != target[-1]):
        print(-1)
    else:
        print(answer)


if __name__ == "__main__":
    main()

File: sections/63_atcoder_difficulty_ladder/problems/41_median_pyramid_hard/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    string start, target;
    if (!(cin >> n >> start >> target)) return 0;

    int previous_operation = 0;
    int answer = 0;
    for (int index = 0; index + 1 < n; index++) {
        int mismatch = start[index] != target[index];
        int current_operation = mismatch ^ previous_operation;
        answer += current_operation;
        previous_operation = current_operation;
    }

    if (previous_operation != (start[n - 1] != target[n - 1])) cout << -1 << '\n';
    else cout << answer << '\n';
    return 0;
}
